libs/test-proxy-master/src/main.rs

```diff
@@ -1,68 +1,121 @@
+
     let addr_clients = SocketAddr::from(([127, 0, 0, 1], 4000));
     let addr_proxies = SocketAddr::from(([127, 0, 0, 1], 5000));
-    let (client_sender, client_receiver): (Sender<Vec<u8>>, Receiver<Vec<u8>>) =
-        broadcast::channel(16);
-    let (proxy_sender, proxy_receiver): (Sender<Vec<u8>>, Receiver<Vec<u8>>) =
-        broadcast::channel(16);
 
     let client_listener = TcpListener::bind(addr_clients).await?;
     println!("Listening on http://{}", addr_clients);
     let proxy_listener = TcpListener::bind(addr_proxies).await?;
     println!("Listening on http://{}", addr_proxies);
 
+    let pool1 = pool.clone();
     let task_1 = tokio::task::spawn(async move {
-        connect_streams_via_channel(
-            client_listener,
-            client_sender,
-            proxy_receiver,
-            Context::Client,
-        )
-        .await
+        while let Ok((stream, _addr)) = proxy_listener.accept().await {
+            let pool = pool1.clone();
+            if let Err(err) = server::conn::http1::Builder::new()
+                .preserve_header_case(true)
+                .title_case_headers(true)
+                .serve_connection(
+                    TokioIo::new(stream),
+                    service_fn(move |req| proxy_endpoint(pool.clone(), req)),
+                )
+                .with_upgrades()
+                .await
+            {
+                println!("Failed to serve connection: {:?}", err);
+            }
+        }
     });
 
     let task_2 = tokio::task::spawn(async move {
-        connect_streams_via_channel(
-            proxy_listener,
-            proxy_sender,
-            client_receiver,
-            Context::Proxy,
-        )
-        .await
+        while let Ok((mut stream, _addr)) = client_listener.accept().await {
+            let pool = pool.clone();
+            tokio::spawn(async move {
+                let mut buf = Vec::new();
+
+                loop {
+                    let mut headers = [httparse::EMPTY_HEADER; 64];
+                    let mut req = httparse::Request::new(&mut headers);
+
+                    let n = stream.read_buf(&mut buf).await?;
+
+                    if n == 0 {
+                        panic!("invalid request")
+                    }
+
+                    if let Ok(httparse::Status::Complete(len)) = req.parse(&buf) {
+                        // We can pick proxy by req
+                        let mut proxy = TokioIo::new(pool.get().await.unwrap());
+
+                        // We can modify the request here
+                        proxy.write_all(&buf[..len]).await?;
+
+                        // Write the rest of the buffer
+                        proxy.write_all(&buf[len..]).await?;
+                        tokio::io::copy_bidirectional(&mut stream, &mut proxy).await?;
+                        break;
+                    }
+                }
+
+                Ok::<(), std::io::Error>(())
+            });
+        }
```

Comment:
I don't like this part.
Because it uses a kinda manual parsing here.
But this looks unavoidable for now

Reply:
Yeah I agree, I'm trying to play with it and see if something else can be used.